app.go

```diff
@@ -261,22 +262,67 @@ func (t stopTimeoutOption) String() string {
 	return fmt.Sprintf("fx.StopTimeout(%v)", time.Duration(t))
 }
 
+// withLogger will stay private until we export fxlog.Logger once
+// we finalize the API.
+func withLogger(l fxlog.Logger) Option {
+	return withLoggerOption{l}
+}
+
+type withLoggerOption struct{ l fxlog.Logger }
+
+func (l withLoggerOption) apply(app *App) {
+	app.log = l.l
+}
+
+func (l withLoggerOption) String() string {
+	return fmt.Sprintf("fx.withLogger(%v)", l.l)
+}
+
 // Printer is the interface required by Fx's logging backend. It's implemented
 // by most loggers, including the one bundled with the standard library.
+//
+// Note, this will be deprecate with next release and you will need to implement
+// fxlog.Logger interface instead.
 type Printer interface {
 	Printf(string, ...interface{})
 }
 
 // Logger redirects the application's log output to the provided printer.
+//
+// Note, this will be deprecated with the next release and you will need to use
+// withLogger instead.
 func Logger(p Printer) Option {
 	return loggerOption{p}
 }
 
 type loggerOption struct{ p Printer }
 
+type printerWrapper struct {
+	p Printer
+}
+
+// newPrinter returns an implementation of zapcore.WriteSyncer
+// used to support Logger option which implements Printer
+// interface.
+func newPrinter(p Printer) zapcore.WriteSyncer {
+	return &printerWrapper{
+		p: p,
+	}
+}
+
+func (p *printerWrapper) Write(b []byte) (n int, err error) {
+	p.p.Printf(string(b))
+	return len(b), nil
+}
+
+func (p *printerWrapper) Sync() error {
+	return nil
+}
+
 func (l loggerOption) apply(app *App) {
-	app.logger = &fxlog.Logger{Printer: l.p}
-	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.logger)}
+	np := newPrinter(l.p)
+	app.log = fxlog.DefaultLogger(np) // assuming np is thread-safe.
+	app.lifecycle = &lifecycleWrapper{lifecycle.New(app.log)}
```

Comment:
Instead of setting up the lifecycle here as well as on line 482, why not set the lifecycle after applying options.

```
app := &App{
  // ...
}

for _, opt := options { opt.apply(app) }

app.lifecycle = ...
```